lalrpop/src/normalize/token_check/mod.rs

```diff
@@ -21,19 +21,50 @@ mod test;
 
 pub fn validate(mut grammar: Grammar) -> NormResult<Grammar> {
     let (has_enum_token, all_literals) = {
+        let opt_match_token = grammar.match_token();
+
+        let (match_mappings, match_catch_all) = if let Some(mt) = opt_match_token {
+            let mut ret = map();
+            let mut catch_all = false;
+
+            // FIXME: This should probably move _inside_ the Validator
+            // FIXME: Add precedences here
+            for mc in &mt.contents {
+                for item in &mc.items {
+                    // TODO: Mabye move this into MatchItem methods
+                    match *item {
+                        MatchItem::Unmapped(sym, _)        => { ret.insert(sym, sym); },
+                        MatchItem::Mapped(sym, mapping, _) => { ret.insert(mapping, sym); },
+                        MatchItem::CatchAll(_)             => { catch_all = true; }
+                    };
+                }
+            }
+
+            (Some(ret), Some(catch_all))
+        } else {
+            (None, None)
+        };
+
         let opt_enum_token = grammar.enum_token();
         let conversions = opt_enum_token.map(|et| {
             et.conversions.iter()
                           .map(|conversion| conversion.from)
-                          .collect()
+                          .collect::<Set<TerminalString>>()
```

Comment:
(Usually, I put such annotations on the local variables themselves, though in this case this might be a better choice, since the line with the local variable would get quite long.

Example:

```rust
let conversions: Option<Set<_>> = opt_enum_token.map(|et| {
    ...
};
```